File: dotnet/DesignPatterns/BehavioralDesignPatterns/ObserverDesignPattern/ObserverDesignPattern.Example1/Author.cs
namespace ObserverDesignPattern.Example1;

public class Author : ISubject<Article>
{
    private readonly List<IObserver<Article>> _followers = new();
    public string Name { get; }

    public Author(string name)
    {
        Name = name;
    }

    public void Subscribe(IObserver<Article> observer)
    {
        if (!_followers.Contains(observer))
            _followers.Add(observer);
    }

    public void Unsubscribe(IObserver<Article> observer)
    {
        _followers.Remove(observer);
    }

    public void Notify(Article article)
    {
        foreach (var follower in _followers)
        {
            follower.Update(article);
        }
    }

    public void PublishArticle(string title, string content)
    {
        var article = new Article(title, content);
        Console.WriteLine($"{Name} published a new article: {title}");
        Notify(article);
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/ObserverDesignPattern/ObserverDesignPattern.Example1/src/Author.ts
import { ISubject } from './ISubject';
import { IObserver } from './IObserver';
import { Article } from './Article';

export class Author implements ISubject<Article> {
  readonly name: string;
  private readonly _followers: IObserver<Article>[] = [];

  constructor(name: string) {
    this.name = name;
  }

  subscribe(observer: IObserver<Article>): void {
    if (!this._followers.includes(observer)) {
      this._followers.push(observer);
    }
  }

  unsubscribe(observer: IObserver<Article>): void {
    const index = this._followers.indexOf(observer);
    if (index !== -1) {
      this._followers.splice(index, 1);
    }
  }

  notify(article: Article): void {
    for (const follower of this._followers) {
      follower.update(article);
    }
  }

  publishArticle(title: string, content: string): void {
    const article = new Article(title, content);
    console.log(`${this.name} published a new article: ${title}`);
    this.notify(article);
  }
}
